practices/appeng/adrs/0002-distributed-tracing.md

```diff
@@ -0,0 +1,157 @@
+are:
+
+- radically improves developer productivity and output
+- works across multiple applications, programming languages, and transports
+- improve time to market
+- facilitates excellent cross-team communication and cooperation
+
+Here are several example scenarios or questions that distributed tracing can
+help answer:
+
+- As a new engineer on the team, I want to understand how many separate systems
+  are involved when a certain user type logs in and the first page is rendered.
+- As an operations engineer, I want to know how many SQL queries are executed
+  for a given endpoint or interaction.
+- As a product manager, I want to know if a new feature is being used by a
+  certain cohort of users on a regular basis.
+- As an engineer, I want to prove that an optimization I wrote is effective
+  in a production environment.
+- As a load tester, after I have shown that a problem exists, I want to
+  understand how the system is interacting so I can debug and fix the issue.
+
+### ADR Goals and Anti-goals
+
+- Goal: Choose which set of libraries to use within a codebase (across
+  programming languages) for generating distributed tracing data
+- Anti-goal: Committing to a specific "backend", i.e. platform or service for
+  gathering, exploring, and displaying trace information
+  - Sub-goal: Leaving open as many options as possible for the backend
+
+## Decision
+
+### Use OpenTelemetry
+
+- OpenTelemetry is an emerging industry standard
+  - vendors find benefit of being in the OpenTelemetry ecosystem because they
+    no longer have to create or support instrumentation libraries in an ever
+    growing array of languages, i.e. as soon as language library exists for
+    OpenTelemetry, the vendors automatically become available to support that
+    given language.
+- OpenTelemetry is vendor agnostic
+  - tracing information can be sent to hosted services (e.g. Honeycomb.io, AWS
+    X-Ray, etc) or self-hosted Open Source implementations (e.g. Zipkin, Jaeger,
+    etc)
+  - if left unconfigured, OpenTelemetry instrumentation calls default to
+    lightweight/noop executions
+- OpenTelemetry has well-maintained libraries for the languages used in the
+  layers of most Truss projects
+  - i.e. Go (back-end); JavaScript (front-end); Python (load testing); etc
+- Easily swappable back-ends
+  - e.g. could choose a local Docker version of OpenZipkin for an all-local
+    development environment
+  - e.g. can use Honeycomb.io in the experimental commercial-cloud hosted
+    environment
+  - e.g. can swap in AWS X-Ray for use in GovCloud hosted environments
+- Cons
+  - as an abstraction layer, OpenTelemetry may prohibit usage of vendor-specific
+    capabilities
+  - some OpenTelemetry libraries and tools may trail their vendor-supported
+    counterparts
+  - instrumentation for tracing may be a vector for performance overhead
```

Comment:
I am not sure how I feel about this line here. I don't know if it's accurate or not. My understanding of observability / OpenTel is that the metrics collected and shipped would actually help with performance overhead and not contribute to it.

Is this line suggesting that adding instrumentation could cause the app to slow down? Or is it suggestion that developer performance could slow down due to adding instrumentation?

```suggestion
  - instrumentation for tracing may be a vector for developer-performance overhead because of additional work to implement into roadmap and product planning
```

Reply:
I think it may be in reference to the fact that in order to run otel, we are in fact running another service and the main application will need to be passing metrics, traces, etc. to the collector so that can impact performance. There's things we can do to modify how much it affects, e.g. change what percentage of things get sampled, but regardless, there will be some impact.

Reply:
My interpretation matched what @felipe-lee said.

Reply:
I don't really agree here. The OpenTel specification addresses this. I think it's too general to say that it could impact performance. The client and SDK authors take this into account. I think it's up to the users of these tools to verify, but not be concerned with the performance issues.

[=> OpenTel Spec on performance](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/performance.md)

Reply:
I suggest we leave this as is for now -- I think there's always the potential that adding another service/dependency/network hop could affect performance, particularly if not integrated or configured properly.  Perhaps we can cover specifics about any performance impacts (or lack of) in the playbook in a "best practices" section (which I expect we will be building as a follow-on to this ADR).

Reply:
I don't feel like what I'm trying to say is coming across appropriately. But I'll try again. The performance that this is implying is that it's affecting the application. I'm not sure that's actually the original concern here. The performance could very well happen on the OpenTel side, the collector, the parser, the aggregator of the data. But the ADR is clear that without a backend the OpenTel calls would be _lightweight_ and a _noop_ which to me are the opposite of _a vector for performance overhead_. 

But like I said in my original review, this is not a blocker, just wanted to make sure it's captured and expressed somewhere. 👍🏽 I wasn't around for the original ~ADR~ MDR being written, but I would have had the same concern then too. The spec is very clear that this is a concern of the authors of OpenTel tools and not the consumers of those tools who are being targeted with this MDR/ADR.